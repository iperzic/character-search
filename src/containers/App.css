.App {
    text-align: center;
    font-family: Arial, Helvetica Neue, Helvetica, sans-serif;
    padding-bottom: 10px;
}
