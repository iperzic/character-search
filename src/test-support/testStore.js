class TestStore {
  static createDefaultAppStore() {
    return ({
      searchValue: '',
      error: '',
    });
  }

  static createQueriedAppStore(value) {
    return ({
      searchValue: value,
      error: '',
    });
  }

  static createErroredAppStore(error) {
    return ({
      searchValue: '',
      error,
    });
  }

  static createMetadataStore(offset = 0, count = 0, total = 0) {
    return ({
      count,
      offset,
      total,
    });
  }
}

export default TestStore;
