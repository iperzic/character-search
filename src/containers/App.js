import React, { Component } from 'react';
import { connect } from 'react-redux';

import './App.css';

import SearchBar from '../components/SearchBar';
import CharactersGrid from '../components/CharactersGrid';
import Pagination from '../components/Pagination';
import Loading from '../components/Loading';

import * as appActions from '../actions/app';

const mapStateToProps = state => ({
  characters: state.characters.search.map(c => ({
    ...state.characters.byId[c],
    isBookmark: state.characters.bookmarks.includes(c),
  })),
  bookmarks: state.characters.bookmarks.map(c => ({
    ...state.characters.byId[c],
    isBookmark: true,
  })),
  searchValue: state.app.searchValue,
  error: state.app.error,
  metadata: state.metadata,
  loading: state.app.loading,
});

const mapDispatchToProps = dispatch => ({
  searchCharacter: value => dispatch(appActions.searchCharacter(value)),
  toggleCharacterBookmark: id => dispatch(appActions.toggleCharacterBookmark(id)),
  changePage: offset => dispatch(appActions.changePage(offset)),
});

class App extends Component {
  constructor(props) {
    super(props);

    this.handleQueryChange = this.handleQueryChange.bind(this);
    this.handleToggleBookmark = this.handleToggleBookmark.bind(this);
    this.handlePageChange = this.handlePageChange.bind(this);
  }

  handleQueryChange(event) {
    const value = event.target.value;
    this.props.searchCharacter(value);
  }

  handleToggleBookmark(id) {
    this.props.toggleCharacterBookmark(id);
  }

  handlePageChange(offset) {
    this.props.changePage(offset);
  }

  render() {
    const isLoading = this.props.loading;
    const shouldPaginationDisplay = this.props.searchValue && !isLoading;
    return (
      <div className="App">
        {isLoading && <Loading />}
        <SearchBar onQueryChange={this.handleQueryChange} />
        {!isLoading && <CharactersGrid
          characters={this.props.searchValue ? this.props.characters : this.props.bookmarks}
          onToggleBookmark={this.handleToggleBookmark}
        />}
        {shouldPaginationDisplay && <Pagination metadata={this.props.metadata} onPageChange={this.handlePageChange} />}
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(App);
