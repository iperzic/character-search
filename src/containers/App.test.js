import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';

import App from './App';
import testStore from '../test-support/testStore';

describe('App', () => {
  let store;

  describe('when the there is no search query', () => {

    describe('and there is no bookmarks', () => {
      beforeEach(() => {
        store = configureStore()(testStore.createFreshStore());
      });

      it('should show empty search input and empty grid', () => {
        const appContainer = shallow(<App store={store} />).dive();

        expect(appContainer.getElement()).toMatchSnapshot();
      });
    });

    describe('and bookmarks exist', () => {
      beforeEach(() => {
        const id = 123;
        const byId = { [id]: { id, name: 'Deadpool', image: 'some-image.jpg' } };
        const characters = testStore.createCharacterStore(byId, undefined, [id]);
        const metadata = testStore.createMetadataStore();
        const app = testStore.createAppStore();
        store = configureStore()({ characters, metadata, app });
      });

      it('should show empty search input and grid with bookmarks', () => {
        const appContainer = shallow(<App store={store} />).dive();

        expect(appContainer.getElement()).toMatchSnapshot();
      });
    });
  });

  describe('when the search query is set', () => {
    beforeEach(() => {
      const id = 123;
      const byId = { [id]: { id, name: 'Deadpool', image: 'some-image.jpg' } };
      const characters = testStore.createCharacterStore(byId, [id]);
      const metadata = testStore.createMetadataStore();
      const app = testStore.createAppStore('Deadpool');
      store = configureStore()({ characters, metadata, app });
    });

    it('should show search input and grid with fetched characters', () => {
      const appContainer = shallow(<App store={store} />).dive();

      expect(appContainer.getElement()).toMatchSnapshot();
    });
  });

  describe('when the data is being fetched', () => {
    beforeEach(() => {
      const characters = testStore.createCharacterStore();
      const metadata = testStore.createMetadataStore();
      const app = testStore.createAppStore(undefined, undefined, true);
      store = configureStore()({ characters, metadata, app });
    });

    it('should show loading screen', () => {
      const appContainer = shallow(<App store={store} />).dive();

      expect(appContainer.getElement()).toMatchSnapshot();
    });
  })
});
