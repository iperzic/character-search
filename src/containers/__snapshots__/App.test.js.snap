// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App when the data is being fetched should show loading screen 1`] = `
<div
  className="App"
>
  <Loading />
  <SearchBar
    onQueryChange={[Function]}
  />
  
</div>
`;

exports[`App when the search query is set should show search input and grid with fetched characters 1`] = `
<div
  className="App"
>
  <SearchBar
    onQueryChange={[Function]}
  />
  <CharactersGrid
    characters={
      Array [
        Object {
          "id": 123,
          "image": "some-image.jpg",
          "isBookmark": false,
          "name": "Deadpool",
        },
      ]
    }
    onToggleBookmark={[Function]}
  />
  <Pagination
    metadata={
      Object {
        "count": 0,
        "offset": 0,
        "total": 0,
      }
    }
    onPageChange={[Function]}
  />
</div>
`;

exports[`App when the there is no search query and bookmarks exist should show empty search input and grid with bookmarks 1`] = `
<div
  className="App"
>
  <SearchBar
    onQueryChange={[Function]}
  />
  <CharactersGrid
    characters={
      Array [
        Object {
          "id": 123,
          "image": "some-image.jpg",
          "isBookmark": true,
          "name": "Deadpool",
        },
      ]
    }
    onToggleBookmark={[Function]}
  />
  
</div>
`;

exports[`App when the there is no search query and there is no bookmarks should show empty search input and empty grid 1`] = `
<div
  className="App"
>
  <SearchBar
    onQueryChange={[Function]}
  />
  <CharactersGrid
    characters={Array []}
    onToggleBookmark={[Function]}
  />
  
</div>
`;
